### ISSUE TITLE
[BUG] Incomplete throttling circumvention

### ISSUE THREAD

**Issue by windows-server-2003** [pytube/pytube#1072]
I apologize for opening an issue without testing but only by reading the source code due to #1060.

**Nature of the issue**
In #1037, the proper calculation of the `n` parameter was implemented, which I really appreciate.
However, I found some issues in the code.

1.  
    https://github.com/pytube/pytube/blob/master/pytube/extract.py#L464
    Due to this `continue`, videos without ciphered signatures will not have the correct `n`.  
    It seems that `n` modification is needed whether the signature is obfuscated or not.
  
2.
    In this line, we are excluding the streams with `ratebypass=yes` in the url parameters from `n` modification.
    https://github.com/pytube/pytube/blob/master/pytube/extract.py#L472
    It seems that the streams with `ratebypass=yes` are the ones in ytInitialPlayerResponse["streamingData"]["formats"] instead of ytInitialPlayerResponse["streamingData"]["adaptiveFormats"] and they have a different `n` value from ones in "adaptiveFormats" from the first. However, as far as I tested with another YouTube parser, these streams also need the same `n` modification to avoid the throttling.

I'm sorry if I'm misunderstanding the code or other things.

**Comment by 89z** [comment]
example video?

**Comment by tfdahlin** [comment]
> 
> 
>     1. https://github.com/pytube/pytube/blob/master/pytube/extract.py#L464
>        Due to this `continue`, videos without ciphered signatures will not have the correct `n`.
>        It seems that `n` modification is needed whether the signature is obfuscated or not.
> 

You're absolutely right about that. That's a mistake on my end, I will fix this after the 404 issue gets fixed.

>     2. In this line, we are excluding the streams with `ratebypass=yes` in the url parameters from `n` modification.
>        https://github.com/pytube/pytube/blob/master/pytube/extract.py#L472
>        It seems that the streams with `ratebypass=yes` are the ones in ytInitialPlayerResponse["streamingData"]["formats"] instead of ytInitialPlayerResponse["streamingData"]["adaptiveFormats"] and they have a different `n` value from ones in "adaptiveFormats" from the first. However, as far as I tested with another YouTube parser, these streams also need the same `n` modification to avoid the throttling.

Good catch. When I first saw this, I suspected that the streams with ratebypass had a pre-calculated value of n (we've seen this before with stream URLs vs streamCiphers), but if those values need to be changed as well, I can take care of that. I didn't extensively test the rate limiting issue because I didn't see it myself, so I wasn't sure if the fix I wrote was complete.

> 
> 
> I'm sorry if I'm misunderstanding the code or other things.

You're good! You caught a couple of mistakes, and I'll be happy to fix them when I have the time to do so c:

**Comment by windows-server-2003** [comment]
Thanks!
I remembered I had one more thing to report :
https://github.com/pytube/pytube/blob/master/pytube/cipher.py#L521
The `h` value is hardcoded here, but it's different at least in the following base.js (`0...9a...zA...Z-_` instead of `A...Za...z0...9-_`
https://m.youtube.com/s/player/02486e7d/player_ias.vflset/ja_JP/base.js (function name : `kha`)
The safer solution is to also emulate the `for` + `switch` part.

**Comment by tfdahlin** [comment]
> 
> 
> Thanks!
> I remembered I had one more thing to report :
> https://github.com/pytube/pytube/blob/master/pytube/cipher.py#L521
> The `h` value is hardcoded here, but it's different at least in the following base.js (`0...9a...zA...Z-_` instead of `A...Za...z0...9-_`
> https://m.youtube.com/s/player/02486e7d/player_ias.vflset/ja_JP/base.js (function name : `kha`)
> The safer solution is to also emulate the `for` + `switch` part.

Hmmmmm. I'll look into emulating it, but it may be a little tricky. In theory, nobody should see the `ja_JP` base.js because of the request headers we use, but i'll look into what it would take to emulate

**Comment by windows-server-2003** [comment]
That's true, though we have to monitor because they may give us such base.js in the future.

**Comment by tfdahlin** [comment]
100% possible that they will, but I'm also not doing true emulation right now, so I need to watch out for that anyway. Currently, pytube effectively has 6 functions that it maps regex matches to. If there is a breaking change like that, the value of `h` is just as likely as them adding a new function type I think.

**Comment by windows-server-2003** [comment]
Because my parser failed, I noticed another thing...
Integers in the `c` array sometimes use the exponential notation like `3289E3` (=3289000), which Python's `int()` can't handle.
Test url (which is what I get when I access with the same `User-Agent` and `Accept-Language` as pytube right now) : 
https://www.youtube.com/s/player/4aeb5fe3/player_ias.vflset/en_US/base.js
There's `113585E4` inside `c`.
